src/main/java/io/nats/client/Connection.java

```diff
@@ -234,10 +234,45 @@
      */
     public void flush(Duration timeout) throws TimeoutException, InterruptedException;
 
+    /**
+     * Drain tells the connection to process in flight messages before closing.
+     * 
+     * Drain initially drains all of the consumers, stopping incoming messages.
+     * Next, publishing is halted and a flush call is used to insure all published
+     * messages have reached the server.
+     * Finally the connection is closed.
+     * 
+     * In order to drain subscribers, an unsub is sent to the server followed by a flush.
+     * These two steps occur before drain returns. The remaining steps occur in a background thread.
+     * This method tries to manage the timeout properly, so that if the timeout is 1 second, and the flush
+     * takes 100ms, the remaining steps have 900ms in the background thread.
+     * 
+     * The connection will try to let all messages be drained, but when the timeout is reached
+     * the connection is closed and any outstanding dispatcher threads are interrupted.
+     * 
+     * A future is used to allow this call to be treated as synchronous or asynchronous as
+     * needed by the application. The value of the future will be true if all of the subscriptions
+     * were drained in the timeout, and false otherwise. The future is completed after the connection
+     * is closed, so any connection handler notifications will happen before the future completes.
+     * 
+     * @param timeout The time to wait for the drain to succeed, pass 0 to wait

```

Comment:
In other languages we use -1 for "forever".

Reply:
since we don't support 0 time, i think this is ok. Also, changing would require a major revision bump.